### PR TITLE
Master decouple staging: fixes up for endpoint destruction

### PR DESCRIPTION
**Pull request by arnop2**

### CODE REVIEW

File: lib/rpmsg/rpmsg.c [hunk 1]
```diff
@@ -264,7 +264,8 @@ void rpmsg_destroy_ept(struct rpmsg_endpoint *ept)
 		return;
 
 	rdev = ept->rdev;
-	(void)rpmsg_send_ns_message(ept, RPMSG_NS_DESTROY);
+	if(ept->name && strncmp(ept->name, "NS", sizeof("NS")))
```
Comment (wjliang) [inline lib/rpmsg/rpmsg.c]: instead of compare the name, how about check  the endpoint address

File: lib/rpmsg/rpmsg_virtio.c [hunk 2]
```diff
@@ -478,10 +478,11 @@ static int rpmsg_virtio_ns_callback(struct rpmsg_endpoint *ept, void *data,
 	_ept = rpmsg_get_endpoint(rdev, name, RPMSG_ADDR_ANY, dest);
 
 	if (ns_msg->flags & RPMSG_NS_DESTROY) {
-		if (_ept) {
-			metal_mutex_release(&rdev->lock);
-			rpmsg_destroy_ept(_ept);
-		}
+		if (_ept)
+			_ept->dest_addr = RPMSG_ADDR_ANY;
```
Comment (wjliang) [inline lib/rpmsg/rpmsg_virtio.c]: why setting the dest_addr to ANY?

File: lib/rpmsg/rpmsg_virtio.c [hunk 3]
```diff
@@ -478,10 +478,11 @@ static int rpmsg_virtio_ns_callback(struct rpmsg_endpoint *ept, void *data,
...
+		metal_mutex_release(&rdev->lock);
+		if (_ept && _ept->destroy_cb)
+				_ept->destroy_cb(ept);
```
Comment (wjliang) [inline lib/rpmsg/rpmsg_virtio.c]: the destroyed callback is to used to notify application the endpoint has been destroyed, better to update the name service bind endpoint callback to also handle destroy callback by adding argument to indicate if it is destroy or announcement. but the rpmsg_virtio_ns_callback should still have default behavior if no such callback is defined by application.

File: lib/rpmsg/rpmsg_virtio.c (hunk 2 above)
Comment (arnop2) [reply lib/rpmsg/rpmsg_virtio.c]: if application don't destroy the endpoint, this allows to reconnect the end point on a new  ns announcement ( for instance crash usecase)

File: lib/rpmsg/rpmsg.c (hunk 1 above)
Comment (arnop2) [reply lib/rpmsg/rpmsg.c]: yes will change this

File: lib/rpmsg/rpmsg_virtio.c (hunk 3 above)
Comment (arnop2) [reply lib/rpmsg/rpmsg_virtio.c]: i will do a patch in this way

Comment (arnop2) [reply lib/rpmsg/rpmsg_virtio.c]: I reanalyses the point. We have following use cases:
- we receive NS_DESTROY message:  
  => call a unbind call back to inform application that the remote endpoint is destroyed. apllication can take appropriate action (suspend communication for this service, destroy the local endpoint...)
- the local endpoint is destroyed. 
=> only use case is the rpmsg_deinit_vdev call. In this case the rpmsg dev is released. I think there is no reason to call a destroy callback during the deinit; application should ensure that all the communications are stopped before calling the denit . So we could just clean the callback call in  rpmsg_deinit_vdev.

In conclusion, i will propose a patch that :
- rename "destroy_cb" in "unbind_cb"
- suppress the callback call in rpmsg_deinit_vdev.

File: lib/rpmsg/rpmsg.c [hunk 4]
```diff
@@ -150,24 +150,26 @@ struct rpmsg_endpoint *rpmsg_get_endpoint(struct rpmsg_device *rdev,
...
 			name_match = !strncmp(ept->name, name,
 					      sizeof(ept->name));
-		if (dest_addr == RPMSG_ADDR_ANY &&
-		    ept->addr == RPMSG_ADDR_ANY &&
-		    name_match)
-			return ept;
-		if (addr == RPMSG_ADDR_ANY &&
-		    ept->dest_addr == RPMSG_ADDR_ANY &&
-		    name && name_match)
+		if (!name || !name_match)
```
Comment (wjliang) [inline lib/rpmsg/rpmsg.c]: looks like name_match may be used as not initialized. As in the previous line name_match is only set when name is not NULL.

File: lib/include/openamp/rpmsg.h [hunk 5]
```diff
@@ -72,7 +73,7 @@ struct rpmsg_endpoint {
 	uint32_t addr;
 	uint32_t dest_addr;
 	rpmsg_ept_cb cb;
-	rpmsg_ept_destroy_cb destroy_cb;
+	rpmsg_ns_unbind_cb unbind_cb;
```
Comment (wjliang) [inline lib/include/openamp/rpmsg.h]: I think the destroy callback is used in the destroy endpoint function call, as the destroy endpoint function can be called internally, it will be good to have a callback to notify user the endpoint has been destroyed, the "unbind" endpoint is used for ns destroy name service. I think it can be merged into the rpmsg_ns_bind_cb maybe rename rpmsg_ns_bind_cb to rpmsg_ns_cb()?

File: lib/rpmsg/rpmsg.c (hunk 4 above)
Comment (arnop2) [reply lib/rpmsg/rpmsg.c]: No issue as name is tested first.  so test here is equivalent to
if( !name) 
     continue
else if (!name_match)
     continue

File: lib/include/openamp/rpmsg.h (hunk 5 above)
Comment (arnop2) [reply lib/include/openamp/rpmsg.h]: Please have a look to the commit message, from my POV destroy_cb is not useful.
Furthermore  we need a callback per endpoint for the unbind, not a global call back, has endpoints can be handled by different users.

File: lib/rpmsg/rpmsg.c [hunk 6]
```diff
@@ -268,6 +268,4 @@ void rpmsg_destroy_ept(struct rpmsg_endpoint *ept)
 	metal_mutex_acquire(&rdev->lock);
 	rpmsg_unregister_endpoint(ept);
 	metal_mutex_release(&rdev->lock);
-	if (ept->destroy_cb)
-		ept->destroy_cb(ept);
 }
```
[the remark this answers is not in the document]
Comment (arnop2) [reply lib/rpmsg/rpmsg.c]: Could you share a use case where the callback is needed? From my POV  this function is called in 2 cases:
1) by user, to destroy the end point,  cb not needed in this case.
2) user calls rpmsg_deinit_vdev, in this case application should free all its resources after the function call, no need of a specific callback.

File: lib/rpmsg/rpmsg.c (hunk 4 above)
Comment (wjliang) [reply lib/rpmsg/rpmsg.c]: you're right. Should be fine. Just not sure if there will be compilation warning. Will know when I compile with the patch.

File: lib/include/openamp/rpmsg.h (hunk 5 above)
Comment (wjliang) [reply lib/include/openamp/rpmsg.h]: if replace unbind_cb, what should we do with destroy_endpoint()? and who should call destroy_endpoint?
in the destroy_endpoint, we should call unbind_cb() first? and then unregister the endpoint if user want to unregister?
How does user application descides when to remove the endpoint or keep it?
or always assume the unbind_cb() is only called from NS destroy? in this case, should the callback name be more clearer so that user knows it is only called when the NS destroy is received?

Comment (arnop2) [reply lib/include/openamp/rpmsg.h]: >if replace unbind_cb, what should we do with destroy_endpoint()? and who should call >destroy_endpoint?
rpmsg_destroy_ept should be called by application when stop the service, to unregister the local end point and send associated NS_DESTROY message.  

>in the destroy_endpoint, we should call unbind_cb() first? and then unregister the endpoint if >user want to unregister?
No, unbind_cb is used only to inform application that remote end point is destroyed. So unbind_cb() can call rpmsg_destroy_ept to desrtroy the local one and free the associated memory ( dynamic allocation case).

>How does user application descides when to remove the endpoint or keep it?
>or always assume the unbind_cb() is only called from NS destroy?
Exactly, unbind_cb() only called on NS_DESTROY to inform that local endpoint is no more bound to the remote one( that is destroyed)

> in this case, should the callback name be more clearer so that user knows it is only called when the NS destroy is received?
What about "ns_unbind_cb" that is symetric with the "ns_bind_cb" defined in  rpmsg_device struct?

File: lib/rpmsg/rpmsg.c (hunk 4 above)
Comment (wjliang) [reply lib/rpmsg/rpmsg.c]: when I compile it i did got warning saying that it may be used uninitialized, i will change to assign default value to 0 when declaring the variable.

File: apps/examples/echo/rpmsg-ping.c [hunk 7]
```diff
@@ -49,7 +49,7 @@ static int rpmsg_endpoint_cb(struct rpmsg_endpoint *ept, void *data, size_t len,
 	(void)ept;
 	(void)src;
 	(void)priv;
-	LPRINTF(" received payload number %lu of size %d \r\n",
+	LPRINTF(" received payload number %lu of size %ld \r\n",
```
Comment (wjliang) [inline apps/examples/echo/rpmsg-ping.c]: it doesn't work for 32bit system. i will make the change.

**Comment by wjliang** [comment]
merged with modification
